### PR TITLE
[Cheah Jing Feng] iP

### PR DESCRIPTION
**Pull request by cloudy3**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -6,5 +9,91 @@ public static void main(String[] args) {
...
+                int dividerPosition = phrase.indexOf("/");
+                String eventTime = phrase.substring(dividerPosition+4);
+                phrase = phrase.substring(6, dividerPosition);
+                Event event = new Event(phrase, eventTime);
+                System.out.println("Got it! I've added this task:");
+                tasks.add(event);
+                System.out.println(event);
+                System.out.println("Now you have " + tasks.size() + " tasks in the list.");
+                System.out.println("~____________________________________________________________~\n");
+            } else {
```
Comment (jhjhajh) [inline src/main/java/Duke.java]: It's great that you included this part of the code to catch any invalid input

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -6,5 +9,91 @@ public static void main(String[] args) {
...
+                // Invalid Task
+                System.out.println("~____________________________________________________________~");
+                System.out.println("Sorry mate I do not understand your request. Please specify task :)");
+                System.out.println("~____________________________________________________________~");
+            }
+        }
```
Comment (heyjinwei) [inline src/main/java/Duke.java]: Consider using switch to accomodate the possibilities

File: src/main/java/Task.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    public String getDescription() {
+        return this.description;
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "\u2713" : " ");  // return tick symbol if done
```
Comment (heyjinwei) [inline src/main/java/Task.java]: Consider using if-else statements for easier readability

File: text-ui-test/input.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+todo borrow book
```
Comment (heyjinwei) [inline text-ui-test/input.txt]: Consider testing with a wider variety of test cases

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -6,5 +9,91 @@ public static void main(String[] args) {
...
+                phrase = phrase.substring(5);
+                Todo todo = new Todo(phrase);
+                System.out.println("Got it! I've added this task:");
+                tasks.add(todo);
+                System.out.println(todo);
+                System.out.println("Now you have " + tasks.size() + " tasks in the list.");
+                System.out.println("~____________________________________________________________~\n");
+            } else if (phrase.startsWith("deadline ")) {
+                // Add a Deadline
+                System.out.println("~____________________________________________________________~");
```
Comment (heyjinwei) [inline src/main/java/Duke.java]: Consider having a method to print this divider or having it as a constant string to print

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -6,5 +9,91 @@ public static void main(String[] args) {
...
+                int dividerPosition = phrase.indexOf("/");
```
Comment (heyjinwei) [inline src/main/java/Duke.java]: Consider getting the indexOf the entire string instead of a single character to avoid unwanted outcomes with inputs with '/'

**Review by heyjinwei — COMMENTED** [review]
Overall the code seems to abide by the coding standards and has a good code quality. However, alternative methods can be explored to improve readability of the code

File: src/main/java/Duke.java (hunk 2 above)
Comment (cloudy3) [reply src/main/java/Duke.java]: Great idea, thanks!

File: src/main/java/Duke.java (hunk 5 above)
Comment (cloudy3) [reply src/main/java/Duke.java]: Good idea!

File: text-ui-test/input.txt (hunk 4 above)
Comment (jhjhajh) [inline text-ui-test/input.txt]: You can also consider testing with corner cases, such as _null_ or try inputs that don't fit your command format

File: src/main/java/Duke.java [hunk 7]
```diff
@@ -1,10 +1,112 @@
+import java.util.ArrayList;
```
Comment (jhjhajh) [inline src/main/java/Duke.java:13]: overall the code adheres to the coding guidelines, but there are some minor changes that can be made to improve on code readability, as suggested by the other reviewer.

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -1,10 +1,214 @@
...
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        DukePrint.printLogo();
+
+        Scanner sc = new Scanner(System.in);
```
Comment (rabhijit) [inline src/main/java/Duke.java]: In general, it's best to use English words as variable names that makes it easier to tell what the variable is, especially later down in the code

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,214 @@
...
+            // Set a task as done
+            try {
+                int taskIndex = phrase.charAt(phrase.length() - 1) - '0';
+
+                Task task = tasks.get(taskIndex - 1);
+                task.markAsDone();
+                System.out.println("Nice! I've marked this task as done:");
+                System.out.println("[" + task.getStatusIcon() + "] " + task.getDescription());
+            } catch (NumberFormatException e) {
+                System.out.println("\uD83D\uDE2D Please input a valid number!");
```
Comment (rabhijit) [inline src/main/java/Duke.java]: Perhaps you could save your unicode as a static final constant, to make it easier to see what you're printing

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,214 @@
...
+            break;
+        default:
+            System.out.println("Invalid data!");
+            break;
+        }
+    }
+
+    public static void saveFile() throws IOException {
+        File path = new File("tasks.txt");
+        if (!path.exists()) {
```
Comment (rabhijit) [inline src/main/java/Duke.java]: Consider merging these into a single boolean condition if possible; it's best to avoid unnecessary nesting

File: src/main/java/models/Deadline.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package models;
+
+public class Deadline extends Task {
+
+    protected String by;
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
```
Comment (rabhijit) [inline src/main/java/models/Deadline.java:14]: You could use more detailed variable names, that could help a reader immediately see what your variable is supposed to represent. Don't be afraid to use slightly lengthier names for clarity! I noticed that some of your variable names are rather brief.

**Review by rabhijit — COMMENTED** [review]
Overall, great work! Your code is really neat and a breeze to read through. You avoided deeply nested code and mostly stuck to a consistent layer of abstraction throughout. One improvement you could make, is that you could use more detailed variable names and perhaps slightly more in-depth comments to explain what your code is doing.
